File: packages/python/elizaos/bootstrap/actions/roles.py
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import TYPE_CHECKING
from uuid import UUID

from elizaos.bootstrap.utils.xml import parse_key_value_xml
from elizaos.types import Action, ActionExample, ActionResult, Content, ModelType

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State


class Role(str, Enum):
    OWNER = "OWNER"
    ADMIN = "ADMIN"
    MEMBER = "MEMBER"
    GUEST = "GUEST"
    NONE = "NONE"


UPDATE_ROLE_TEMPLATE = """# Update entity role in the world.

{{providers}}

# Current Role Assignments:
{{#each roles}}
- {{this.entityId}}: {{this.role}}
{{/each}}

Valid roles: OWNER, ADMIN, MEMBER, GUEST, NONE

<response>
    <thought>Your reasoning for the role change</thought>
    <entity_id>The entity ID to update</entity_id>
    <new_role>The new role to assign (OWNER, ADMIN, MEMBER, GUEST, or NONE)</new_role>
</response>"""


@dataclass
class UpdateRoleAction:
    name: str = "UPDATE_ROLE"
    similes: list[str] = field(
        default_factory=lambda: [
            "ASSIGN_ROLE",
            "CHANGE_ROLE",
            "SET_ROLE",
            "MODIFY_PERMISSIONS",
            "GRANT_ROLE",
        ]
    )
    description: str = (
        "Update the role of an entity in a world. Use this to manage permissions and access levels."
    )

    async def validate(
        self, runtime: IAgentRuntime, message: Memory, _state: State | None = None
    ) -> bool:
        room_id = message.room_id
        if not room_id:
            return False

        room = await runtime.get_room(room_id)
        if room is None or room.world_id is None:
            return False

        world = await runtime.get_world(room.world_id)
        if world is None or world.metadata is None:
            return False

        roles = world.metadata.get("roles", {})
        agent_role = roles.get(str(runtime.agent_id), Role.NONE.value)

        return agent_role in (Role.OWNER.value, Role.ADMIN.value)

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        if state is None:
            raise ValueError("State is required for UPDATE_ROLE action")

        room_id = message.room_id
        if not room_id:
            return ActionResult(
                text="No room context for role update",
                values={"success": False, "error": "no_room_id"},
                data={"actionName": "UPDATE_ROLE"},
                success=False,
            )

        room = await runtime.get_room(room_id)
        if room is None or room.world_id is None:
            return ActionResult(
                text="Room or world not found",
                values={"success": False, "error": "room_not_found"},
                data={"actionName": "UPDATE_ROLE"},
                success=False,
            )

        world = await runtime.get_world(room.world_id)
        if world is None or world.metadata is None:
            return ActionResult(
                text="World not found",
                values={"success": False, "error": "world_not_found"},
                data={"actionName": "UPDATE_ROLE"},
                success=False,
            )

        state = await runtime.compose_state(
            message, ["RECENT_MESSAGES", "ACTION_STATE", "WORLD_INFO"]
        )

        current_roles = world.metadata.get("roles", {})
        roles_context = "\n".join(
            f"- {entity_id}: {role}" for entity_id, role in current_roles.items()
        )

        template = (
            runtime.character.templates.get("updateRoleTemplate")
            if runtime.character.templates and "updateRoleTemplate" in runtime.character.templates
            else UPDATE_ROLE_TEMPLATE
        )
        prompt = runtime.compose_prompt(state=state, template=template)
        prompt = prompt.replace("{{roles}}", roles_context)

        response_text = await runtime.use_model(ModelType.TEXT_LARGE, prompt=prompt)
        parsed_xml = parse_key_value_xml(response_text)

        if parsed_xml is None:
            raise ValueError("Failed to parse XML response")

        thought = str(parsed_xml.get("thought", ""))
        entity_id_str = str(parsed_xml.get("entity_id", ""))
        new_role_str = str(parsed_xml.get("new_role", "")).upper()

        if not entity_id_str:
            raise ValueError("No entity ID provided")

        if new_role_str not in [r.value for r in Role]:
            raise ValueError(f"Invalid role: {new_role_str}")

        entity_id = UUID(entity_id_str)

        roles = dict(world.metadata.get("roles", {}))
        old_role = roles.get(str(entity_id), Role.NONE.value)
        roles[str(entity_id)] = new_role_str
        world.metadata["roles"] = roles

        await runtime.update_world(world)

        response_content = Content(
            text=f"Updated role for {entity_id_str}: {old_role} -> {new_role_str}",
            actions=["UPDATE_ROLE"],
        )

        if callback:
            await callback(response_content)

        return ActionResult(
            text=f"Role updated: {entity_id_str} is now {new_role_str}",
            values={
                "success": True,
                "roleUpdated": True,
                "entityId": str(entity_id),
                "oldRole": old_role,
                "newRole": new_role_str,
            },
            data={
                "actionName": "UPDATE_ROLE",
                "entityId": str(entity_id),
                "oldRole": old_role,
                "newRole": new_role_str,
                "thought": thought,
            },
            success=True,
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return [
            [
                ActionExample(
                    name="{{name1}}",
                    content=Content(text="Make @user an admin in this server."),
                ),
                ActionExample(
                    name="{{name2}}",
                    content=Content(
                        text="I'll update their role to admin.",
                        actions=["UPDATE_ROLE"],
                    ),
                ),
            ],
        ]


update_role_action = Action(
    name=UpdateRoleAction.name,
    similes=UpdateRoleAction().similes,
    description=UpdateRoleAction.description,
    validate=UpdateRoleAction().validate,
    handler=UpdateRoleAction().handler,
    examples=UpdateRoleAction().examples,
)

File: packages/rust/src/bootstrap/actions/roles.rs
//! UPDATE_ROLE action implementation.

use async_trait::async_trait;
use std::sync::Arc;
use uuid::Uuid;

use crate::error::{PluginError, PluginResult};
use crate::runtime::{IAgentRuntime, ModelParams};
use crate::types::{ActionResult, Memory, ModelType, Role, State};
use crate::xml::parse_key_value_xml;

use super::Action;

const UPDATE_ROLE_TEMPLATE: &str = r#"# Task: Update entity role in the world.

{{providers}}

# Current Role Assignments:
{{roles}}

# Instructions:
Based on the request, determine the role assignment to make.
Valid roles are: OWNER, ADMIN, MEMBER, GUEST, NONE

Respond using XML format like this:
<response>
    <thought>Your reasoning for the role change</thought>
    <entity_id>The entity ID to update</entity_id>
    <new_role>The new role to assign (OWNER, ADMIN, MEMBER, GUEST, or NONE)</new_role>
</response>

IMPORTANT: Your response must ONLY contain the <response></response> XML block above."#;

/// Action for updating entity roles.
pub struct UpdateRoleAction;

#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
impl Action for UpdateRoleAction {
    fn name(&self) -> &'static str {
        "UPDATE_ROLE"
    }

    fn similes(&self) -> &[&'static str] {
        &["ASSIGN_ROLE", "CHANGE_ROLE", "SET_ROLE", "MODIFY_PERMISSIONS", "GRANT_ROLE"]
    }

    fn description(&self) -> &'static str {
        "Update the role of an entity in a world. \
         Use this to manage permissions and access levels."
    }

    async fn validate(&self, runtime: &dyn IAgentRuntime, message: &Memory) -> bool {
        let Some(room_id) = message.room_id else {
            return false;
        };

        // Check if room exists and has a world
        let Ok(Some(room)) = runtime.get_room(room_id).await else {
            return false;
        };

        let Some(world_id) = room.world_id else {
            return false;
        };

        // Check if agent has permission to update roles
        let Ok(Some(world)) = runtime.get_world(world_id).await else {
            return false;
        };

        if let Some(roles) = world.metadata.get("roles") {
            if let Some(roles_obj) = roles.as_object() {
                let agent_id = runtime.agent_id().to_string();
                if let Some(role) = roles_obj.get(&agent_id) {
                    let role_str = role.as_str().unwrap_or("NONE");
                    return role_str == "OWNER" || role_str == "ADMIN";
                }
            }
        }

        false
    }

    async fn handler(
        &self,
        runtime: Arc<dyn IAgentRuntime>,
        message: &Memory,
        state: Option<&State>,
        _responses: Option<&[Memory]>,
    ) -> PluginResult<ActionResult> {
        let _state = state.ok_or_else(|| {
            PluginError::StateRequired("State is required for UPDATE_ROLE action".to_string())
        })?;

        let room_id = message.room_id.ok_or_else(|| {
            PluginError::InvalidInput("No room context for role update".to_string())
        })?;

        let room = runtime
            .get_room(room_id)
            .await?
            .ok_or_else(|| PluginError::NotFound("Room not found".to_string()))?;

        let world_id = room.world_id.ok_or_else(|| {
            PluginError::InvalidInput("Room has no world".to_string())
        })?;

        let world = runtime
            .get_world(world_id)
            .await?
            .ok_or_else(|| PluginError::NotFound("World not found".to_string()))?;

        // Compose state
        let composed_state = runtime
            .compose_state(message, &["RECENT_MESSAGES", "ACTION_STATE", "WORLD_INFO"])
            .await?;

        // Get template
        let template = runtime
            .character()
            .templates
            .get("updateRoleTemplate")
            .map(|s| s.as_str())
            .unwrap_or(UPDATE_ROLE_TEMPLATE);

        let prompt = runtime.compose_prompt(&composed_state, template);

        // Call the model
        let response = runtime
            .use_model(ModelType::TextLarge, ModelParams::with_prompt(&prompt))
            .await
            .map_err(|e| PluginError::ModelError(e.to_string()))?;

        let response_text = response
            .as_text()
            .ok_or_else(|| PluginError::ModelError("Expected text response".to_string()))?;

        // Parse XML response
        let parsed = parse_key_value_xml(response_text)
            .ok_or_else(|| PluginError::XmlParse("Failed to parse response XML".to_string()))?;

        let thought = parsed.get("thought").cloned().unwrap_or_default();
        let entity_id_str = parsed
            .get("entity_id")
            .cloned()
            .ok_or_else(|| PluginError::InvalidInput("No entity ID provided".to_string()))?;
        let new_role_str = parsed
            .get("new_role")
            .cloned()
            .ok_or_else(|| PluginError::InvalidInput("No role provided".to_string()))?
            .to_uppercase();

        // Validate entity ID
        let entity_id = Uuid::parse_str(&entity_id_str).map_err(|_| {
            PluginError::InvalidInput(format!("Invalid entity ID: {}", entity_id_str))
        })?;

        // Validate role
        let new_role: Role = match new_role_str.as_str() {
            "OWNER" => Role::Owner,
            "ADMIN" => Role::Admin,
            "MEMBER" => Role::Member,
            "GUEST" => Role::Guest,
            "NONE" => Role::None,
            _ => return Err(PluginError::InvalidInput(format!("Invalid role: {}", new_role_str))),
        };

        // Get old role
        let old_role = world
            .metadata
            .get("roles")
            .and_then(|r| r.as_object())
            .and_then(|o| o.get(&entity_id_str))
            .and_then(|r| r.as_str())
            .unwrap_or("NONE")
            .to_string();

        // Update role in world
        let mut updated_world = world.clone();
        let roles = updated_world
            .metadata
            .entry("roles".to_string())
            .or_insert_with(|| serde_json::json!({}));
        if let Some(roles_obj) = roles.as_object_mut() {
            roles_obj.insert(entity_id_str.clone(), serde_json::json!(new_role_str));
        }
        runtime.update_world(&updated_world).await?;

        Ok(ActionResult::success(format!(
            "Role updated: {} is now {}",
            entity_id_str, new_role_str
        ))
        .with_value("success", true)
        .with_value("roleUpdated", true)
        .with_value("entityId", entity_id_str.clone())
        .with_value("oldRole", old_role.clone())
        .with_value("newRole", new_role_str.clone())
        .with_data("actionName", "UPDATE_ROLE")
        .with_data("entityId", entity_id_str)
        .with_data("oldRole", old_role)
        .with_data("newRole", new_role_str)
        .with_data("thought", thought))
    }
}

